### PR TITLE
Feature/implement refund

### PR DESCRIPTION
**Pull request by Alexey-N-Chernyshov**
add effective refund implementation and tests

### CODE REVIEW

**Comment by Alexey-N-Chernyshov** [comment]
I have to update bindings and irohalib fro linux.

File: src/main/kotlin/endpoint/RefundServerEndpoint.kt [hunk 1]
```diff
@@ -26,16 +26,14 @@ class RefundServerEndpoint(
...
     private val ethNotaryAdapter = moshi.adapter(EthNotaryResponse::class.java)!!
 
     init {
         val server = embeddedServer(Netty, port = serverBundle.port) {
             routing {
-                get(serverBundle.ethRefund) {
-                    logger.info { "EthRefund invoked" }
-                    logger.info { "parameters:${call.parameters}" }
-                    call.respondText { onCallEthRefund(call.parameters["contract"]) }
+                get(serverBundle.ethRefund + "/{tx_hash}") {
```
Comment (x3medima17) [inline src/main/kotlin/endpoint/RefundServerEndpoint.kt]: You can use $tx_hash

File: src/main/kotlin/endpoint/RefundServerEndpoint.kt [hunk 2]
```diff
@@ -58,10 +55,18 @@ class RefundServerEndpoint(
             })
     }
 
+    /**
+     * Creates a [EthRefundRequest] object from request string
+     */
+    private fun createRequest(txHash: String?): EthRefundRequest? {
+        // TODO replace destination ethereum address
+        return if (txHash != null) EthRefundRequest(txHash) else null
```
Comment (x3medima17) [inline src/main/kotlin/endpoint/RefundServerEndpoint.kt]: Use Elvis operator

File: src/main/kotlin/endpoint/RefundServerEndpoint.kt [hunk 3]
```diff
@@ -58,10 +55,18 @@ class RefundServerEndpoint(
...
+    }
+
     /**
      * Method return response on stateless invalid request
      */
-    fun onErrorPipelineCall(): String {
+    private fun onErrorPipelineCall(): String {
         logger.error { "Request has been failed" }
         return ""
```
Comment (x3medima17) [inline src/main/kotlin/endpoint/RefundServerEndpoint.kt]: I don't think it's a good idea to return empty string

File: src/main/kotlin/endpoint/eth/EthNotaryResponse.kt [hunk 4]
```diff
@@ -68,7 +65,7 @@ class EthNotaryResponseMoshiAdapter {
     @FromJson
     fun fromJson(layer: EthNotaryResponseLayer) = when (layer.type) {
         EthNotaryResponseType.Successful -> EthNotaryResponse.Successful(layer.ethSignature!!, layer.ethRefund!!)
-        EthNotaryResponseType.Error -> EthNotaryResponse.Error(layer.code!!, layer.reason!!)
+        EthNotaryResponseType.Error -> EthNotaryResponse.Error(layer.reason!!)
```
Comment (x3medima17) [inline src/main/kotlin/endpoint/eth/EthNotaryResponse.kt]: Are you sure reason is always not null?

File: src/main/kotlin/endpoint/eth/EthRefundStrategyImpl.kt [hunk 5]
```diff
@@ -2,70 +2,142 @@ package endpoint.eth
...
-        // rollback case
-        if (commands.hasSetAccountDetail()) {
-            TODO("rollback case")
-        }
+    private fun checkTransaction(
+        appearedTx: iroha.protocol.BlockOuterClass.Transaction,
+        request: EthRefundRequest
+    ): Result<EthRefund, Exception> {
+        return Result.of {
+            val commands = appearedTx.payload.getCommands(0)
```
Comment (x3medima17) [inline src/main/kotlin/endpoint/eth/EthRefundStrategyImpl.kt]: You can use Commands.fromProto()

File: src/main/kotlin/endpoint/eth/EthRefundStrategyImpl.kt [hunk 6]
```diff
@@ -2,70 +2,142 @@ package endpoint.eth
...
 
-        // Iroha -> Eth case
-        if (commands.hasTransferAsset()) {
-            TODO("Withdraw")
-        }
+            when {
+            // rollback case
+                (appearedTx.payload.commandsCount == 1) &&
```
Comment (x3medima17) [inline src/main/kotlin/endpoint/eth/EthRefundStrategyImpl.kt]: Why do you need round brackets?

File: src/test/integration/IntegrationTest.kt [hunk 7]
```diff
@@ -70,26 +98,75 @@ class IntegrationTest {
...
-            }
+    fun sendTxToIroha(txBuilder: ModelTransactionBuilder): String {
+        val utx = txBuilder.build()
+        val hash = utx.hash().hex()
+
+        // sign transaction and get its binary representation (Blob)
+        val txblob = ModelProtoTransaction(utx).signAndAddSignature(keypair).finish().blob().toByteArray()
+
+        // create proto object
+        var protoTx: BlockOuterClass.Transaction? = null
```
Comment (x3medima17) [inline src/test/integration/IntegrationTest.kt]: I would suggest to add some util/ package to minimise code needed to deal with iroha API (in future)

File: src/main/kotlin/sideChain/iroha/util/Uint256Ext.kt [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package sideChain.iroha.util
+
+import iroha.protocol.Primitive.uint256
+import java.math.BigInteger
+
+/**
+ * Extension function that converts [iroha.protocol.Primitive.uint256] to [BigInteger].
+ */
+fun uint256.toBigInteger() = BigInteger.valueOf(this.first).shiftLeft(256)
```
Comment (x3medima17) [inline src/main/kotlin/sideChain/iroha/util/Uint256Ext.kt]: We need a test for that

File: src/main/kotlin/endpoint/RefundServerEndpoint.kt (hunk 1 above)
Comment (Alexey-N-Chernyshov) [reply src/main/kotlin/endpoint/RefundServerEndpoint.kt]: It is a Ktor syntax to obtain parameter from URI.

File: src/main/kotlin/endpoint/RefundServerEndpoint.kt (hunk 2 above)
Comment (Alexey-N-Chernyshov) [reply src/main/kotlin/endpoint/RefundServerEndpoint.kt]: yes, but how to call a method with elvis operator?

File: src/main/kotlin/endpoint/RefundServerEndpoint.kt (hunk 3 above)
Comment (Alexey-N-Chernyshov) [reply src/main/kotlin/endpoint/RefundServerEndpoint.kt]: fixed

File: src/main/kotlin/endpoint/eth/EthNotaryResponse.kt (hunk 4 above)
Comment (Alexey-N-Chernyshov) [reply src/main/kotlin/endpoint/eth/EthNotaryResponse.kt]: Yes, it is for boilerplate code for serialisation.

File: src/test/integration/IntegrationTest.kt (hunk 7 above)
Comment (Alexey-N-Chernyshov) [reply src/test/integration/IntegrationTest.kt]: Yes, but now we have more prioritised tasks to do.

File: src/main/kotlin/endpoint/eth/EthRefundStrategyImpl.kt (hunk 5 above)
Comment (Alexey-N-Chernyshov) [reply src/main/kotlin/endpoint/eth/EthRefundStrategyImpl.kt]: I added task to Jira to move proto out of logic.

File: src/main/kotlin/sideChain/iroha/util/Uint256Ext.kt (hunk 8 above)
Comment (Alexey-N-Chernyshov) [reply src/main/kotlin/sideChain/iroha/util/Uint256Ext.kt]: done

**Pull request merged**